image-cropper/src/app/cropper/canvas/canvas.rs

```diff
@@ -0,0 +1,336 @@
+                },
+                _ => (),
+            }
+        }
+        self.tool.update(event);
+    }
+
+    pub fn render(&self, props: CanvasProps) -> RenderingTree {
+        let image_size = self.image.size();
+        let offset = self.offset.clone();
+        let scale = self.scale.clone();
+        let current_tool_type = self.tool.get_current_tool_type().clone();
+        let hand_tool_pushed_position = self.hand_tool_pushed_position.clone();
+
+        let scaled_image_size = Wh {
+            width: image_size.width * self.scale,
+            height: image_size.height * self.scale,
+        };
+        let scaled_offset = Xy {
+            x: offset.x * self.scale,
+            y: offset.y * self.scale,
+        };
+        render([
+            render_background(&props.wh).attach_event(|builder| {
+                builder
+                    .on_wheel(move |event| {
+                        let offset = offset.clone();
+                        let scale = scale.clone();
+                        let canvas_wh = props.wh.clone();
+                        let image_size = image_size.clone();
+
+                        let managers = namui::managers();
+                        let mouse_manager = &managers.mouse_manager;
+                        let mouse_position = mouse_manager.mouse_position();
+                        let canvas_xy = event
+                            .namui_context
+                            .get_rendering_tree_xy(event.target)
+                            .expect("failed to get canvas xy");
+                        let local_mouse_position = Xy {
+                            x: mouse_position.x as f32 - canvas_xy.x,
+                            y: mouse_position.y as f32 - canvas_xy.y,
+                        };
+                        let is_mouse_in_canvas = !(local_mouse_position.x < 0.0
+                            || local_mouse_position.x > canvas_wh.width
+                            || local_mouse_position.y < 0.0
+                            || local_mouse_position.y > canvas_wh.height);
+
+                        if !is_mouse_in_canvas {
+                            return;
+                        }
+
+                        let keyboard_manager = &managers.keyboard_manager;
+                        if keyboard_manager.any_code_press([namui::Code::ControlLeft]) {
+                            zoom(
+                                event.delta_xy,
+                                &offset,
+                                &local_mouse_position,
+                                &canvas_wh,
+                                &image_size,
+                                scale,
+                            )
+                        } else if keyboard_manager.any_code_press([namui::Code::ShiftLeft]) {
+                            scroll(
+                                &Xy {
+                                    x: event.delta_xy.y,
+                                    y: event.delta_xy.x,
```

Comment:
Is it intended? x=y, y=x?

Reply:
Yes, if we scroll while pressing shift key, canvas will be scrolled with delta x, y changed.
Vertical scroll will be horizontal scroll.